PXC-4003: wsrep_sync_wait has no effect after COM_CHANGE_USER

venkatesh-prasad-v opened the pull request:
https://jira.percona.com/browse/PXC-4003

Problem
-------
wsrep_sync_wait checks are skipped after COM_CHANGE_USER.

Analysis
--------

1. During execution of COM_CHANGE_USER, server calls
   thd->cleanup_connection() in order to clear the context of the
   previous user.
2. thd->cleanup_connection() calls THD::cleanup()
3. THD::cleanup() resets wsrep_client_thread
4. Any query after COM_CHANGE_USER will have the
   thd->wsrep_client_thread set to false.
5. These queries skip wsrep_sync_waits because these checks are done
   only when wsrep_client_thread = true

Fix
---
Move the cleanup of wsrep_client_thread to end_connection() instead of
THD::cleanup_connection().

Note
---
 - This bug is not reproducible on 5.7.


Testing done
----

Without fix
```
$ php -f wsrep_sync_wait---com_change_user.php.txt | sort | uniq -c | sort -k 3,4

     14 LOOP2: node2  count: 19 - sum: 171 - wsrep_sync_wait: 15
     19 LOOP3: node2  count: 18 - sum: 153 - wsrep_sync_wait: 7
      1 LOOP2: node2  count: 11 - sum: 55 - wsrep_sync_wait: 15
      1 LOOP2: node2  count: 4 - sum: 6 - wsrep_sync_wait: 15
      1 LOOP2: node2  count: 7 - sum: 21 - wsrep_sync_wait: 15
      1 LOOP2: node2  count: 8 - sum: 28 - wsrep_sync_wait: 15
      1 LOOP3: node2  count: 11 - sum: 55 - wsrep_sync_wait: 7
      1 LOOP3: node2  count: 12 - sum: 66 - wsrep_sync_wait: 7
     21 LOOP3: node2  count: 19 - sum: 171 - wsrep_sync_wait: 7
      2 LOOP2: node2  count: 13 - sum: 78 - wsrep_sync_wait: 15
      2 LOOP2: node2  count: 15 - sum: 105 - wsrep_sync_wait: 15
      2 LOOP2: node2  count: 16 - sum: 120 - wsrep_sync_wait: 15
      2 LOOP3: node2  count: 13 - sum: 78 - wsrep_sync_wait: 7
      2 LOOP3: node2  count: 15 - sum: 105 - wsrep_sync_wait: 7
    351 LOOP3: node2  count: 20 - sum: 190 - wsrep_sync_wait: 7
    367 LOOP2: node2  count: 20 - sum: 190 - wsrep_sync_wait: 15
      3 LOOP2: node2  count: 10 - sum: 45 - wsrep_sync_wait: 15
      3 LOOP2: node2  count: 14 - sum: 91 - wsrep_sync_wait: 15
      3 LOOP3: node2  count: 14 - sum: 91 - wsrep_sync_wait: 7
      4 LOOP3: node2  count: 16 - sum: 120 - wsrep_sync_wait: 7
    500 LOOP1: node2  count: 20 - sum: 210 - wsrep_sync_wait: 15
      7 LOOP2: node2  count: 17 - sum: 136 - wsrep_sync_wait: 15
      7 LOOP2: node2  count: 18 - sum: 153 - wsrep_sync_wait: 15
     87 LOOP3: node2  count: 20 - sum: 210 - wsrep_sync_wait: 7
     89 LOOP2: node2  count: 20 - sum: 210 - wsrep_sync_wait: 15
      9 LOOP3: node2  count: 17 - sum: 136 - wsrep_sync_wait: 7
     14 LOOP3: node3  count: 18 - sum: 153 - wsrep_sync_wait: 7
     15 LOOP2: node3  count: 19 - sum: 171 - wsrep_sync_wait: 15
      1 LOOP2: node3  count: 10 - sum: 45 - wsrep_sync_wait: 15
      1 LOOP2: node3  count: 8 - sum: 28 - wsrep_sync_wait: 15
      1 LOOP3: node3  count: 10 - sum: 45 - wsrep_sync_wait: 7
      1 LOOP3: node3  count: 12 - sum: 66 - wsrep_sync_wait: 7
      1 LOOP3: node3  count: 14 - sum: 91 - wsrep_sync_wait: 7
      1 LOOP3: node3  count: 2 - sum: 1 - wsrep_sync_wait: 7
      1 LOOP3: node3  count: 6 - sum: 15 - wsrep_sync_wait: 7
      1 LOOP3: node3  count: 7 - sum: 21 - wsrep_sync_wait: 7
     20 LOOP3: node3  count: 19 - sum: 171 - wsrep_sync_wait: 7
      2 LOOP2: node3  count: 12 - sum: 66 - wsrep_sync_wait: 15
      2 LOOP2: node3  count: 13 - sum: 78 - wsrep_sync_wait: 15
    352 LOOP2: node3  count: 20 - sum: 190 - wsrep_sync_wait: 15
    353 LOOP3: node3  count: 20 - sum: 190 - wsrep_sync_wait: 7
      3 LOOP3: node3  count: 13 - sum: 78 - wsrep_sync_wait: 7
      3 LOOP3: node3  count: 15 - sum: 105 - wsrep_sync_wait: 7
      4 LOOP2: node3  count: 14 - sum: 91 - wsrep_sync_wait: 15
      4 LOOP2: node3  count: 16 - sum: 120 - wsrep_sync_wait: 15
    500 LOOP1: node3  count: 20 - sum: 210 - wsrep_sync_wait: 15
      5 LOOP2: node3  count: 15 - sum: 105 - wsrep_sync_wait: 15
      5 LOOP3: node3  count: 17 - sum: 136 - wsrep_sync_wait: 7
      6 LOOP3: node3  count: 16 - sum: 120 - wsrep_sync_wait: 7
     90 LOOP3: node3  count: 20 - sum: 210 - wsrep_sync_wait: 7
     96 LOOP2: node3  count: 20 - sum: 210 - wsrep_sync_wait: 15
      9 LOOP2: node3  count: 17 - sum: 136 - wsrep_sync_wait: 15
      9 LOOP2: node3  count: 18 - sum: 153 - wsrep_sync_wait: 15
```

With fix
```
$ php -f wsrep_sync_wait---com_change_user.php.txt | sort | uniq -c | sort -k 3,4

    500 LOOP1: node2  count: 20 - sum: 210 - wsrep_sync_wait: 15
    500 LOOP2: node2  count: 20 - sum: 210 - wsrep_sync_wait: 15
    500 LOOP3: node2  count: 20 - sum: 210 - wsrep_sync_wait: 7
    500 LOOP1: node3  count: 20 - sum: 210 - wsrep_sync_wait: 15
    500 LOOP2: node3  count: 20 - sum: 210 - wsrep_sync_wait: 15
    500 LOOP3: node3  count: 20 - sum: 210 - wsrep_sync_wait: 7
```

kamil-holubicki commented on sql/sql_parse.cc:
```diff
@@ -2174,6 +2174,7 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
       thd->cleanup_connection();
 
 #ifdef WITH_WSREP
+      thd->wsrep_client_thread = true;
```
cleanup_connection() is called in two cases: COM_CHANGE_USER and COM_RESET_CONNECTION  . Do we have the same problem in the 2nd case as well?
Maybe it would be better to set wsrep_client_thread = true after THD::cleanup() in cleanup_connection() ?

venkatesh-prasad-v replied:
Yes. You are right. COM_RESET_CONNECTION  (\x) is also affected.

``` 
Thread 40 "connection" hit Breakpoint 1, dispatch_command (thd=0x7f0fec002090, com_data=0x7f1004785b40, 
    command=COM_RESET_CONNECTION) at /home/venki/work/pxc/80/sql/sql_parse.cc:2113
2113          wsrep_after_command_ignore_result(thd);
(gdb) n
2114          wsrep_close(thd);
(gdb) n
2116          thd->cleanup_connection();
(gdb) p thd->wsrep_client_thread
$4 = true
(gdb) n
2118          wsrep_open(thd);
(gdb) p thd->wsrep_client_thread
$5 = false
 ```
 

> Maybe it would be better to set wsrep_client_thread = true after THD::cleanup() in cleanup_connection() ? 

IMO removing the `wsrep_client_thread = false` line  from  `THD::cleanup()` and placing inside `end_connection()` would be a better solution. We set `wsrep_client_thread = true` inside `thd_prepare_connection()`. So, moving the cleanup code to inside `end_connection()` would be reasonable in this case.

venkatesh-prasad-v commented:
Jenkins: https://pxc.cd.percona.com/view/8.0%20parallel%20MTR/job/pxc-8.0-pipeline-parallel-mtr/527/consoleFull

Failing tests: None

Commit:
PXC-4003: wsrep_sync_wait has no effect after COM_CHANGE_USER

https://jira.percona.com/browse/PXC-4003

Problem
-------
wsrep_sync_wait checks are skipped after COM_CHANGE_USER.

Analysis
--------

1. During execution of COM_CHANGE_USER, server calls
   thd->cleanup_connection() in order to clear the context of the
   previous user.
2. thd->cleanup_connection() calls THD::cleanup()
3. THD::cleanup() resets wsrep_client_thread
4. Any query after COM_CHANGE_USER will have the
   thd->wsrep_client_thread set to false.
5. These queries skip wsrep_sync_waits because these checks are done
   only when wsrep_client_thread = true

Fix
---
Move the cleanup of wsrep_client_thread to end_connection() instead of
THD::cleanup_connection().

dutow commented on sql/sql_class.cc:
```diff
@@ -1353,7 +1353,6 @@ void THD::cleanup(void) {
   if (wsrep_cs().state() != wsrep::client_state::s_none) {
```
Shouldn't the client state and wsrep_client_thread be handled the same way?

With the client state, it's set to s_none here, and later it's reset to s_idle specifically in sql_parse for these commands, which seems to be a hack.

venkatesh-prasad-v replied:
IMO, wsrep_client_thread and client state are two separate things and need not be handled the same way. While client state changes as and when the statement proceeds (checks are done at statement level), whereas `wsrep_client_thread` is for the entire session irrespective of the wsrep client state. 

By design `wsrep_client_thread` distinguishes client threads with other threads and will be set to true only in connection handling layer as soon as the server detects that there is a new incoming connection. This is independent of wsrep client state.
